lib/Dialect/AIEVec/Transforms/VectorToAIEVecConversions.cpp

```diff
@@ -2801,6 +2801,95 @@ struct LowerVectorContractionOpToAIEVecMatMulPattern
   bool matMoveToAcc;
 };
 
+// Convert a `vector.transpose` op to an `aievec.shuffle` op for AIEml.
+struct LowerVectorTransposeOpToAIEVecShuffleOpPattern
+    : OpConversionPattern<vector::TransposeOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(vector::TransposeOp transpOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto resTy = transpOp.getResultVectorType();
+    auto resShape = resTy.getShape();
+    auto elemTyBitWidth = resTy.getElementTypeBitWidth();
+    auto vBitWidth = std::accumulate(resShape.begin(), resShape.end(),
+                                     elemTyBitWidth, std::multiplies<>());
+    if (vBitWidth != 512)
+      return failure();
+
+    if (elemTyBitWidth != 8 && elemTyBitWidth != 16 && elemTyBitWidth != 32)
+      return failure();
+
+    // Verify leading dimensions are all 1.
+    for (int64_t i = 0; i < static_cast<int64_t>(resShape.size() - 2); ++i)
+      if (resShape[i] != 1)
+        return failure();
+
+    // Only permutation of the 2 innermost dimensions are supported.
+    ArrayRef<int64_t> perm = transpOp.getPermutation();
+    for (int64_t i = 0; i < static_cast<int64_t>(perm.size() - 2); ++i)
+      if (perm[i] != i)
+        return failure();
+    if (perm[perm.size() - 1] != static_cast<int64_t>(perm.size() - 2))
+      return failure();
+
+    auto id = aievec::ShuffleMode::T512_1X2_LO;
+    static const aievec::ShuffleMode T8Modes[17] = {
```

Comment:
Perhaps use a `llvm::densemap` or a `std::unordered_map` here? So that we can get rid of all those default `id`s.

Reply:
Mapping 3-4 values under 17 to another integral value under 256 is a bit of an overkill, maps are slow and tough on caches, so I'd rather use LUTs for this. Alternatively, I could just do if-then-else (it's what I originally had).